### PR TITLE
feat(weave): add typed attribute maps for fast filtering

### PR DESCRIPTION
**Pull request by gtarpenning**
## Summary

Adds four typed Map columns to `call_parts`, `calls_merged`, and `calls_complete` alongside `attributes_dump`. At ingest, each attribute value is routed into `attributes_map_str` / `_int` / `_float` / `_bool` by Python type. Filters wrapped in `$convert` hit the typed column directly instead of `JSON_VALUE(attributes_dump, ...)`.

- ORDER BY and unconverted filters keep the JSON_VALUE path — no behavior change for those.
- Modeled on the `custom_attrs*` pattern in #6567.

## Back-compat

Rows inserted before migration 030 (or before backfill completes) have empty typed maps. The query builder emits a row-level `if(mapContains(...), ...)` guard so the fast path fires only when the map was populated for that row; legacy rows fall back to the existing JSON_VALUE path. Mixed-backfill tables stay correct without a cutoff timestamp or table-wide flag.

Minimal example — `$convert(attributes.retries, int) == 3` on `calls_complete`:

```sql
if(attributes_map_int != NULL,  -- fix me (can break out to top)
   attributes_map_int['retries'],
   toInt64OrNull(coalesce(nullIf(JSON_VALUE(attributes_dump, '$."retries"'), 'null'), '')))
= 3
```

## Test plan

- [x] 10 new unit tests in `test_attributes_map_dispatch.py` (dispatch per cast, nested paths, fallback, extractor edge cases)
- [x] 366 existing `query_builder/` tests pass
- [x] `clickhouse_trace_server_migrator` tests pass

Local benchmarking:

### CODE REVIEW

File: weave/trace_server/calls_query_builder/calls_query_builder.py [hunk 1]
```diff
@@ -2074,6 +2085,48 @@ def _get_multi_value_feedback_field(
     return None
 
 
+def _maybe_attributes_map_filter_sql(
```
Comment (mscavezze-cw) [inline weave/trace_server/calls_query_builder/calls_query_builder.py:2221]: Does this make an assumption that the table in question has the attributes_map column, and further that it is populated for all rows? Should there be handling for partially back-filled tables?

File: weave/trace_server/clickhouse/utilities.py [hunk 2]
```diff
@@ -23,6 +24,11 @@
 
 logger = logging.getLogger(__name__)
 
+# Per-map cap for the typed attribute maps populated by extract_typed_attrs.
+# Applied independently to each of the four maps, so a single call can
+# contribute up to 4 * MAX_TYPED_ATTR_ENTRIES_PER_MAP entries across all maps.
```
Comment (mscavezze-cw) [inline weave/trace_server/clickhouse/utilities.py]: What happens if we exceed that?
- do we fail the insert?
- do we insert a subset of the values?
- do we insert a sentinel value so that we can be sure not to use the fast sql for this row or table?

**Review by mscavezze-cw — COMMENTED** [review]
Cool - I skimmed the PR and asked some high-level questions.

**Comment by gtarpenning** [comment]
lets consider back compat path with a version number instead, at least be able to break out the check for map existence to the top, all cols should be present if any are.

File: weave/trace_server/calls_query_builder/calls_query_builder.py (hunk 1 above)
Comment (gtarpenning) [reply weave/trace_server/calls_query_builder/calls_query_builder.py:2221]: [:robot: automated] Considered a few back-compat gate options; kept the per-filter `mapContains` for now and wanted to leave the trade-offs in the thread instead of the code.

1. **Per-filter `mapContains(map, key)` (current).** Correct for mixed-backfill rows and preserves cross-type cast semantics — `$convert(attributes.enabled, string) = 'true'` on a bool attribute falls back to `JSON_VALUE` and matches the pre-maps behavior exactly. Downside: duplicates the check across every `$convert` filter in a query.
2. **Row-level "any map has content" gate.** `length(map_str) + length(map_int) + length(map_float) + length(map_bool) > 0` would hoist the check to a single per-row expression (all four maps populate together at ingest). Cheaper, but changes semantics on wrong-type casts: the fast branch returns the Map default (`''`/`0`/`false`/`0.0`) for a key that lives in a sibling typed map, where the JSON fallback would cast it successfully. Also misclassifies empty-attrs rows (all maps empty even post-extraction).
3. **Version column (e.g. `attrs_extracted UInt8`).** Same per-row shape as (2), same semantic trade-off, plus a new migration to add + backfill the marker. Cleanest long-term answer once a backfill job exists — probably worth revisiting then.

Picked (1) because correctness > perf during backfill, and CH inlines the common-subexpression `mapContains` + `map[key]` on populated rows so the duplicated work is smaller than it looks in the SQL.

Separately, removed `MAX_TYPED_ATTR_ENTRIES_PER_MAP` in this push — `_strip_large_values` now treats `attributes_dump` + the four `attributes_map_*` columns as one slot, so oversized rows get the existing `ENTITY_TOO_LARGE_PAYLOAD` treatment instead of silent truncation at extract time.

File: weave/trace_server/clickhouse/utilities.py (hunk 2 above)
Comment (gtarpenning) [reply weave/trace_server/clickhouse/utilities.py]: [🤖 automated] Went with option (b) effectively — removed the cap entirely. Oversized rows now fall through to the existing `_strip_large_values` path (which this push also teaches about the `attributes_map_*` columns, so the dump + typed maps get the `ENTITY_TOO_LARGE_PAYLOAD` replacement together). No silent truncation at extract time.

File: tests/trace_server/query_builder/test_attributes_map_dispatch.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+    cq.add_field("id")
+    cq.add_condition(
+        tsi_query.EqOperation.model_validate(
+            {
+                "$eq": [
+                    {
+                        "$convert": {
+                            "input": {"$getField": "attributes.env"},
+                            "to": "int",
+                        }
```
Comment (bcsherma) [inline tests/trace_server/query_builder/test_attributes_map_dispatch.py:44]: Why do we need to do this `$convert`? Since we are comparing to a literal and constrained in general to literal comparisons can't we just infer from the literal type?

File: tests/trace_server/query_builder/test_attributes_map_dispatch.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        cq,
+        """
+        SELECT calls_merged.id AS id
+        FROM calls_merged
+        PREWHERE calls_merged.project_id = {pb_3:String}
+        GROUP BY (calls_merged.project_id, calls_merged.id)
+        HAVING (
+            ((if(mapContains(any(calls_merged.attributes_map_int), {pb_0:String}),
+                 any(calls_merged.attributes_map_int)[{pb_0:String}],
+                 toInt64OrNull(coalesce(nullIf(JSON_VALUE(any(calls_merged.attributes_dump), {pb_1:String}), 'null'), '')))
```
Comment (bcsherma) [inline tests/trace_server/query_builder/test_attributes_map_dispatch.py:62]: So I make sure I understand, we still load the attributes for all rows but we avoid the json parse cost if the typed maps are populated and the key is found

**Review by bcsherma — COMMENTED** [review]
The columns look good, I guess the tricky bit here is backcompat and right now the query is eating that complexity by coalescing new -> old data sources. I am wondering how much this is really going to help bc the attribute dumps are typically pretty small. For inputs/outputs this is much higher value right? How much impact does this have?

File: tests/trace_server/query_builder/test_attributes_map_dispatch.py (hunk 3 above)
Comment (gtarpenning) [reply tests/trace_server/query_builder/test_attributes_map_dispatch.py:44]: Currently all fields are stored as strings in the backend, this is what the frontend does. I do agree that it's possibly not required. I'll investigate.
